### DISCUSSION TITLE
리더보드 사용법에 대해 궁금한 점이 있습니다.

### DISCUSSION THREAD

**Discussion by mayko1234** [naverz/zepeto-studio-kor#303] [5. Q&A]
![image](https://user-images.githubusercontent.com/97988822/151928015-1547007c-74fb-45eb-83e2-cff0c857483b.png)

안녕하세요, 제가 구현하려는 함수는, 리더보드로 부터 유저의 점수를 얻어와서 그 점수를(int) 리턴하는 함수입니다.
GetBestScore( )함수리턴이 OnGetRangeRankResultForBestScore( )함수가 끝나기 전에 이뤄져서 GetBestScore( )의 리턴값이 항상 0입니다. 
제 예상으론 console.log("11" +...) 부분이 console.log("22" +... ) 부분보다 일찍 이뤄져야 할 것 같은데, 실제로는 22가 훨씬 빨리 프린트되고, 11는 그 후에 프린트 됩니다.  

함수가 어떻게 콜 되고 있는 건지(왜 이런 딜레이가 있는 지), 이 부분을 해결하려면 어떻게 코드를 짜는게 좋을 지 조언 부탁드립니다.

**Comment by runasy-koonta** [comment]
Leaderboard API를 써보지 않아서 잘 모르겠지만…ㅠㅠ 아마 콜백이라서 그런 것 같습니다.
서버 호출하는 로직이라 서버 응답 올때까지 대기하면 애플리케이션이 인터럽트 되니 콜백으로 빼서 응답까지 오래 걸리더라도 애플리케이션이 정상적으로 구동되도록 만들어져 있는 것 같습니다.
ZepetoScript에서 Javascript의 Promise가 지원된다면(이 부분은 테스트가 필요할 것 같습니다) Promise를 활용해서 wrapper를 짜신 뒤 async/await을 사용하시면 코드 순서 보장이 될 것 같아요.

다만 저 함수가(GetBestScore) 실행을 방해하면 안 되는 함수라면 (대표적으로 Update 안에서 실행되는 함수가 이에 해당합니다) 콜백에서 최종 작업을 처리하는 것도 고려해보셔야 할 것 같습니다.

**Reply by runasy-koonta** [reply]
테스트해보니 ZepetoScript에서 Promise 사용이 가능하네요.
Promise Wrapper의 예제 코드입니다.

원본 코드가 이렇다면,
![image](https://user-images.githubusercontent.com/49000379/151945004-0e4010d1-a5cc-4b25-a8a6-722d55a0f447.png)
이 상태에서는 2222가 된 후에 1111이 출력되고요,

Promise Wrapper를 사용하면 이렇게 쓸 수 있습니다.
![image](https://user-images.githubusercontent.com/49000379/151944963-5c88d113-0a3a-4b86-a695-d35c80eeccb1.png)
이 상태에서는 1111이 먼저 표시된 후에 2222가 표시됩니다.

**Reply by mayko1234** [reply]
리플 감사합니다. 말씀해주신 방법대로 해봤는데, 제가 아직 자바스크립트에 익숙하지가 않아서 그런지,,, Best Score를 리턴하는 함수를 만들지 못했습니다. return 11111이 async부분이 끝나기 전에 이뤄져서 GetBestScore2( )는 항상 11111만 리턴합니다. 제가 잘못 응용하고 있는 거라면 꼭 알려주세요. ㅠㅠ 외부 클래스에서 LeaderboardManager의 GetBestSocre( )를 콜하면 리더보드에서 얻은 int값만 리턴하면 되는데 쉽지가 않네요.  

![image](https://user-images.githubusercontent.com/97988822/152044446-c13777ee-f465-4756-b221-ac99e95882cc.png)

제가 대안식을 찾아서 원하는 기능 자체는 구현하긴 했는데, 더 좋은 방법이 있지 않을까라는 생각은 떨칠수가 없네요.. 아래 방법을 이용해서 베스트 스코어값을 외부 클래스에서 이용할 수 있었습니다. 

[LeaderboardManager.ts]
![image](https://user-images.githubusercontent.com/97988822/152044823-1022f2cd-dc48-4c2d-be42-2405bf297ab2.png)

[외부 클래스 - GameManager.ts]
![image](https://user-images.githubusercontent.com/97988822/152045128-eb697f18-9d18-4352-a7c5-d4946e40668c.png)

설명도 해주시고 예시 코드까지 공유해주셔서 감사드립니다.

**Comment by x4cf** [comment]
이것도 한번 링크걸어보아요 :D
https://github.com/naverz/zepeto-studio-kor/discussions/137

**Reply by mayko1234** [reply]
이 문제는 클래스의 static instance를 이용해서 해결했어요. 
첨부해주신 솔루션도 후에 참고하겠습니다. 
도움주시려고 리플 달아주셔서 넘 감사합니다.
